zipkin-collector/rabbitmq/src/main/java/zipkin/collector/rabbitmq/RabbitMQCollector.java

```diff
@@ -0,0 +1,202 @@
+    return new Builder();
+  }
+
+  /** Configuration including defaults needed to consume spans from a RabbitMQ queue. */
+  public static final class Builder implements CollectorComponent.Builder {
+    Collector.Builder delegate = Collector.builder(RabbitMQCollector.class);
+    CollectorMetrics metrics = CollectorMetrics.NOOP_METRICS;
+    String queue = "zipkin";
+    ConnectionFactory connectionFactory = new ConnectionFactory();
+    Address[] addresses;
+    int concurrency = 1;
+
+    @Override public Builder storage(StorageComponent storage) {
+      this.delegate.storage(storage);
+      return this;
+    }
+
+    @Override public Builder sampler(CollectorSampler sampler) {
+      this.delegate.sampler(sampler);
+      return this;
+    }
+
+    @Override public Builder metrics(CollectorMetrics metrics) {
+      this.metrics = checkNotNull(metrics, "metrics").forTransport("rabbitmq");
+      this.delegate.metrics(this.metrics);
+      return this;
+    }
+
+    public Builder addresses(List<String> addresses) {
+      this.addresses = convertAddresses(addresses);
+      return this;
+    }
+
+    public Builder concurrency(int concurrency) {
+      this.concurrency = concurrency;
+      return this;
+    }
+
+    public Builder connectionFactory(ConnectionFactory connectionFactory) {
+      this.connectionFactory = checkNotNull(connectionFactory, "connectionFactory");
+      return this;
+    }
+
+    /**
+     * Queue zipkin spans will be consumed from. Defaults to "zipkin-spans".
+     */
+    public Builder queue(String queue) {
+      this.queue = checkNotNull(queue, "queue");
+      return this;
+    }
+
+    @Override public RabbitMQCollector build() {
+      return new RabbitMQCollector(this);
+    }
+  }
+
+  final String queue;
+  final LazyInit connection;
+
+  RabbitMQCollector(Builder builder) {
+    this.queue = builder.queue;
+    this.connection = new LazyInit(builder);
+  }
+
+  @Override
+  public RabbitMQCollector start() {
+    connection.get();
+    return this;
+  }
+
+  @Override
+  public CheckResult check() {
+    try {
+      CheckResult failure = connection.failure.get();
+      if (failure != null) return failure;
+      return CheckResult.OK;
+    } catch (RuntimeException e) {
+      return CheckResult.failed(e);
+    }
+  }
+
+  @Override public void close() throws IOException {
+    connection.close();
+  }
+
+  /** Lazy creates a connection and a queue before starting consumers */
+  static final class LazyInit extends LazyCloseable<Connection> {
+    final Builder builder;
+    final AtomicReference<CheckResult> failure = new AtomicReference<>();
+
+    LazyInit(Builder builder) {
+      this.builder = builder;
+    }
+
+    @Override
+    protected Connection compute() {
+      Connection connection;
+      try {
+        connection = builder.connectionFactory.newConnection(builder.addresses);
+        connection.createChannel().queueDeclare(builder.queue, true, false, false, null);
```

Comment:
The `Channel` created here will remain open until the connection is closed, which is not a big deal but it isn't necessary and it may be asked about by operators looking at open channels on RabbitMQ.

Reply:
I guess this is implicitly closed via the connection, ya?

Reply:
Yes, closing the connection will close any Channels that it opened. Ref. https://www.rabbitmq.com/releases/rabbitmq-java-client/current-javadoc/com/rabbitmq/client/Connection.html#close--